x/mint/abci.go

```diff
@@ -43,3 +56,21 @@ func BeginBlocker(ctx sdk.Context, k Keeper) {
 		),
 	)
 }
+
+// getNewAverageBlockTime calculates the average time it takes to forge a new block based on the cumulated
+// moving average since the beginning of the chain.
+// The formula we use is a cumulated moving average.
+// averageBlockTime + ((currentBlockTime - averageBlockTime) / blockHeight - 1)
+// We remove 1 to blockheight given we dont have an average on block 1.
+func getNewAverageBlockTime(ctx sdk.Context, params types.Params, minter types.Minter) time.Duration {
+	if !minter.LastBlockTimestamp.IsZero() { // Comes from Genesis
+		currentBlockTime := ctx.BlockTime().Sub(minter.LastBlockTimestamp)
+		currentAverage := minter.AverageBlockTime.Nanoseconds() +
+			((currentBlockTime.Nanoseconds() - minter.AverageBlockTime.Nanoseconds()) /
+				(ctx.BlockHeight() - 1))
```

Comment:
why are we subtracting 1?

Reply:
Because there is this issue, imagine on block 1 we don't have a sample of a block time, but on block 2 we have from block1 to block 2 time difference.

But if we use BlockHeight it is 2, so it is not the average because we only have one value.

Reply:
Makes sense.